File: frontend/src/types/pageEditor.ts
/** Page editor types aligned with backend Pydantic models. */

/** Text style for title page elements (text, variable). Mirrors TextOverrideStyles. */
export interface TextStyle {
  font?: string;
  font_size?: number;
  weight?: string;
  style?: string;
  fill?: string;
  tracking?: number;
  word_spacing?: number;
  lang?: string;
  region?: string;
  hyphenate?: boolean | string;
  ligatures?: boolean;
  number_type?: string;
  number_width?: string;
  /** Horizontal alignment for variable elements: left, center, right, justify */
  align?: "left" | "center" | "right" | "justify";
}

export interface Paper {
  name: string;
  width: number;
  height: number;
  margin: number;
}

export interface TextElement {
  type: "text";
  id: string;
  x_mm: number;
  y_mm: number;
  content: string;
  text_style?: TextStyle;
}

export interface VariableElement {
  type: "variable";
  id: string;
  x_mm: number;
  y_mm: number;
  width_mm: number;
  height_lines: number;
  var_name: string;
  text_style?: TextStyle;
}

export interface LineElement {
  type: "line";
  id: string;
  x1_mm: number;
  y1_mm: number;
  x2_mm: number;
  y2_mm: number;
}

export type Element = TextElement | VariableElement | LineElement;

export function isTextElement(e: Element): e is TextElement {
  return e.type === "text";
}
export function isVariableElement(e: Element): e is VariableElement {
  return e.type === "variable";
}
export function isLineElement(e: Element): e is LineElement {
  return e.type === "line";
}

export const PAPER_PRESETS: Record<string, Paper> = {
  a4: { name: "a4", width: 210, height: 297, margin: 20 },
  a3: { name: "a3", width: 297, height: 420, margin: 20 },
  custom: { name: "custom", width: 210, height: 297, margin: 20 },
};

File: app/models/pydantic/page_editor.py
"""Pydantic models for page editor (title pages). Compatible with PIZDoCanvas."""

from typing import Any, Literal, Optional, Union

from pydantic import BaseModel


class PaperSize(BaseModel):
    """Paper dimensions and margin (mm)."""

    name: str = "a4"
    width: float = 210.0
    height: float = 297.0
    margin: float = 20.0


class TextElement(BaseModel):
    """Static text element."""

    type: Literal["text"] = "text"
    id: str
    x_mm: float
    y_mm: float
    content: str = "Sample Text"
    text_style: Optional[dict[str, Any]] = None


class VariableElement(BaseModel):
    """Editable variable with drag + resize."""

    type: Literal["variable"] = "variable"
    id: str
    x_mm: float
    y_mm: float
    width_mm: float = 50.0
    height_lines: float = 2.0
    var_name: str = "var"
    text_style: Optional[dict[str, Any]] = None


class LineElement(BaseModel):
    """Line with endpoints (x1,y1,x2,y2 mm)."""

    type: Literal["line"] = "line"
    id: str
    x1_mm: float
    y1_mm: float
    x2_mm: float
    y2_mm: float


Element = Union[TextElement, VariableElement, LineElement]


class TitlePageContent(BaseModel):
    """Content of a title page: elements, paper, variables."""

    elements: list[Element]
    paper: PaperSize
    variables: dict[str, str] = {}
    ignore_document_styles: bool = True
